compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -549,201 +550,31 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             | ty::InstanceDef::ThreadLocalShim(..)
             | ty::InstanceDef::Item(_) => {
                 // We need MIR for this fn
-                let Some((body, instance)) = M::find_mir_or_eval_fn(
-                    self,
-                    instance,
-                    caller_abi,
-                    args,
-                    destination,
-                    target,
-                    unwind,
-                )?
-                else {
-                    return Ok(());
-                };
-
-                // Compute callee information using the `instance` returned by
-                // `find_mir_or_eval_fn`.
-                // FIXME: for variadic support, do we have to somehow determine callee's extra_args?
-                let callee_fn_abi = self.fn_abi_of_instance(instance, ty::List::empty())?;
-
-                if callee_fn_abi.c_variadic || caller_fn_abi.c_variadic {
-                    throw_unsup_format!("calling a c-variadic function is not supported");
-                }
-
-                if M::enforce_abi(self) {
-                    if caller_fn_abi.conv != callee_fn_abi.conv {
-                        throw_ub_custom!(
-                            fluent::const_eval_incompatible_calling_conventions,
-                            callee_conv = format!("{:?}", callee_fn_abi.conv),
-                            caller_conv = format!("{:?}", caller_fn_abi.conv),
-                        )
+                let body = match M::find_mir_or_extra_fn(self, instance)? {
+                    Either::Left(b) => b,
+                    Either::Right(f) => {
+                        return M::call_extra_fn(
+                            self,
+                            f,
+                            (caller_abi, caller_fn_abi),
+                            args,
+                            destination,
+                            target,
+                            unwind,
+                        );
                     }
-                }
-
-                // Check that all target features required by the callee (i.e., from
-                // the attribute `#[target_feature(enable = ...)]`) are enabled at
-                // compile time.
-                self.check_fn_target_features(instance)?;
-
-                if !callee_fn_abi.can_unwind {
-                    // The callee cannot unwind, so force the `Unreachable` unwind handling.
-                    unwind = mir::UnwindAction::Unreachable;
-                }
+                };
 
-                self.push_stack_frame(
+                self.eval_body(
                     instance,
                     body,
+                    (caller_abi, caller_fn_abi),
+                    args,
+                    with_caller_location,
                     destination,
-                    StackPopCleanup::Goto { ret: target, unwind },
-                )?;
```

Comment:
this can now be moved into the `Left` match arm, unless you want to do the change that returns instances from the `Right` arm under some circumstances

Reply:
dc9b611 👍🏻